### DISCUSSION TITLE
Bulk move of pages, refactoring structure

### DISCUSSION THREAD

**Discussion by fmos** [zim-desktop-wiki/zim-desktop-wiki#2015] [Ideas]
Hi!

I came here to ask whether anyone had an idea on how to perform a bulk move of multiple pages (approximately 1000) from one directory into a subdirectory. The number of pages started to slow down the side pane. I'm switching to a different structure for new pages, but needed to move the old pages.

Moving the pages in the file system and updating the index obviously did not update the links. That alone would not have bothered me too much, but the dead links cause just as many grey entries in the side pane, rendering this approach futile.

Only during writing my question, a relatively simple solution came to mind that I though I'd like to share nevertheless: I ended up renaming the parent directory, individually moving the (much fewer) previous subdirectories into a new parent directory and finally moving the renamed parent as subdirectory into the new parent:

- move "cases" to "unsorted", effectively moving all pages and directories
- create new "cases"
- move directories out of "unsortet" and to new "cases"
- move "unsorted" into new "cases"

Cheers!

**Comment by introt** [comment]
Hi!

I bumped into the `move_page` method while browsing the Notebook API working on the window title stuff. [According to the docs](https://introt.github.io/zim-desktop-wiki/docs/api/zim.notebook.notebook.Notebook.html#move_page), one only needs to call `Notebook.move_page('old:path', 'new:and:improved:path')` to utilize it - meaning it should be fairly easy to expose to the cli via a plugin for scripting.

GUI-wise, while dragging-and-dropping in the index is quite hazardous, we could implement a dedicated treeview widget which generates `(Path('old:existing:one'), Path('prospective:new:one'))` tuples that are sanity-checked and user-confirmed before being moved in bulk.
